neurots/generate/orientations.py

```diff
@@ -360,3 +435,215 @@ def compute_interval_n_tree(soma, n_trees, rng=np.random):
+
+    Args:
+        tmd_parameters (dict): Input parameters.
+        tmd_distributions (dict): Input distributions.
+
+    Returns:
+        tmd_parmeters with fit data if 3d_angles mode is found, else None
+    """
+    with_3d = False
+    morph_class = (
+        "with_apical" if "apical_dendrite" in tmd_parameters["grow_types"] else "without_apical"
+    )
+    new_tmd_parameters = deepcopy(tmd_parameters)
+    for neurite_type in tmd_parameters["grow_types"]:
+
+        if (
+            tmd_parameters[neurite_type] is not None
+            or "mode" not in tmd_parameters[neurite_type]["orientation"]
+        ):
+            continue
+
+        if tmd_parameters[neurite_type]["orientation"]["mode"] == "apical_constraint":
+            with_3d = True
+            if (
+                tmd_parameters[neurite_type]["orientation"].get("values") is None
+                or "params" not in tmd_parameters[neurite_type]["orientation"]["values"]
+            ):
+                data = tmd_distributions[neurite_type]["trunk"]["apical_3d_angles"]
+                new_tmd_parameters[neurite_type]["orientation"]["values"] = _fit_single_3d_angles(
+                    data["data"],
+                    neurite_type,
+                    morph_class,
+                    fit_params=_get_fit_params_from_input_parameters(tmd_parameters[neurite_type]),
+                )
+
+        if tmd_parameters[neurite_type]["orientation"]["mode"] == "pia_constraint":
+            with_3d = True
+            if (
+                tmd_parameters[neurite_type]["orientation"].get("values") is None
+                or "params" not in tmd_parameters[neurite_type]["orientation"]["values"]
+            ):
+                data = tmd_distributions[neurite_type]["trunk"]["pia_3d_angles"]
+                new_tmd_parameters[neurite_type]["orientation"]["values"] = _fit_single_3d_angles(
+                    data["data"],
+                    neurite_type,
+                    morph_class,
+                    fit_params=_get_fit_params_from_input_parameters(tmd_parameters[neurite_type]),
+                )
+    return new_tmd_parameters if with_3d else None
+
+
+def _sample_trunk_from_3d_angle(parameters, rng, tree_type, ref_dir, max_tries=100):
+    """Sample trunk directions from fit of distribution of 3d_angles wrt to ref_dir.
+
+    We use the accept-reject algorithm so we can sample from any distribution.
+    After a number of unsuccessful tries (default=100), we stop and return a random direction.
+    We also issue a warning so the user is aware that the provided distribution may have issues,
+    mostly related to large region of small probabilities.
+    """
+    prob = get_probability_function(
+        form=parameters[tree_type]["orientation"]["values"]["form"],
+        with_density=False,
+    )
+    params = parameters[tree_type]["orientation"]["values"]["params"]
+    n_try = 0
+    while n_try < max_tries:
```

Comment:
so this could be unsuccessful? does this practically happens?

Reply:
yes, if the fit is bad, below there is a warning explaining it